Support different flavours of packages

jacobstanley opened the pull request:
This allows the profiling build or package documentation to be installed
when required, the package is "upgraded" in place in the global cache.

jacobstanley committed:
Support different flavours of packages

This allows the profiling build or package documentation to be installed
when required, the package is "upgraded" in place in the global cache.

charleso commented on src/Mafia/Init.hs:
```diff
@@ -150,18 +154,16 @@ data Profiling =
-
 data MafiaState =
   MafiaState {
-      msInstallState :: InstallState
```
Any reason for this change apart from cleaning up? Looks like the json file has the same information?

jacobstanley replied:
yeah just cleaning up

jacobstanley replied:
whether profiling is installed is morally part of the install state, which is just the mafia state now

bumping the version on the json state is basically free, recreating it only takes 5-10secs

charleso commented on src/Mafia/Install.hs:
```diff
@@ -163,50 +181,96 @@ link w db env p@(Package (PackageRef pid _ _) _ _) = do
+    [ "--reinstall"
+    , "--ghc-options=-fprof-auto-exported"
+    , "--enable-library-profiling" ]
+  Documentation ->
```
I guess at some point we'll want to call this again

jacobstanley replied:
yeah i think we'd call that when someone asks for hoogle or for haddock docs at some point

jacobstanley committed:
Install hackage packages via 'cabal unpack'

Keeping the dist directory around means that we can cheaply upgrade
packages to have profiling enabled.

charleso commented on src/Mafia/Install.hs:
```diff
@@ -163,50 +181,108 @@ link w db env p@(Package (PackageRef pid _ _) _ _) = do
+    -- hackage package
+    Nothing -> do
+      createDirectoryIfMissing True sbsrc
+      Hush <- sbcabal "unpack" ["--destdir=" <> sbsrc, renderPackageId pid]
```
A comment for why you do this might be good - it won't be obvious over time :)

jacobstanley committed:
Add profiling as an option for the chaos test

jacobstanley committed:
Add comment for 'cabal unpack'
